tower_job_template to use organizations

sean-m-sullivan opened the pull request:
##### SUMMARY
Update tower_job_template to use organization similarly to how tower_workflow_job_template does. 

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
 - Collection

##### AWX VERSION
```
awx: 11.2.0
```

##### ADDITIONAL INFORMATION
 - Based on the organization use in tower_workflow_job_template
 - Used to help lookup the project and by extension the job template
 - The Organization is inferred from the associated project
 - If not provided, will lookup by name only, which does not work with duplicates.

 Tested against:
 - Multiple projects with the same name and different organizations
 - Multiple job templates with the same name with different organizations with identically named projects and different inventories (To help tell them apart)
 - Tower 3.6.4
 - Tower 3.7.0
 - AWX 11.2.0

sean-m-sullivan committed:
tower_job_template to use organizations

AlanCoding commented on awx_collection/plugins/modules/tower_job_template.py:
```diff
@@ -415,19 +429,26 @@ def main():
+    organization = module.params.get('organization')
+    if organization:
+        organization_id = module.resolve_name_to_id('organizations', organization)
+        tower_version = module.get_endpoint('ping')['json']['version']
```
I'm generally hesitant to issue an extra request for this purpose, although, we've recently made some changes so that we get it in the response header. Maybe we could develop a shared method which is a performant way to obtain/compare version. @john-westcott-iv I'm all ears if you have ideas for that.

My alternative would be to simply not make this request. Always include `organization` in the JT search payload if given, and document as a backward-incompatible change. This can still be avoided by users with older server versions by not specifying the parameter. Anyway, if they are facing that problem, they aren't matching server-collection versions, which we discourage.

sean-m-sullivan replied:
I'm all for not getting tower version, or just dumping this out and specifying it in docs. 
We could just rely on the user not to supply the org if they are prior to 3.7 and add a note in documentation. 
At the time I was trying to just make sure it would work against 3.7, 3.6 and awx. If that's the consensus, I'll remove the parts that the check version.

john-westcott-iv replied:
We do pull the version out of the headers in module_util/tower_api.py around like 224. It should be fairly trivial to make properties in the object to hold these. On the other side, if we are going to have something that compares version we probably would want to make a method in the tower_api object because you would need to make it work for both tower an AWX.

john-westcott-iv commented:
Unless I am missing the intention of this PR; as a general comment on this, the design pattern we chose for the collection was to go by name or ID only (you can pass the ID in as the name filed and it will find the data). Adding other criteria to lookup secondary objects will simply overburden the modules' parameters. For example, the name of a credential may not be unique and tower_job_template can take multiple credentials so how would you provide additional parameters needed to lookup all the credentials?

Instead, the plan is to provide a tower lookup plugin which will take "arbitrary" criteria and lookup the IDs of objects within Tower. I believe @AlanCoding has started that work. Then, if your installation does not use unique names you would be able to do something akin to:
```
- name: Build a job template
  tower_job_template:
    name: "{{ lookup('tower', {'name': 'My Job Template', 'organization': 'Default'}) }}"
    project: "{{ lookup('tower', {'name': 'My Project', 'organization': 'Default'}) }}"
```

AlanCoding commented:
> Adding other criteria to lookup secondary objects will simply overburden the modules' parameters. For example, the name of a credential may not be unique and tower_job_template can take multiple credentials so how would you provide additional parameters needed to lookup all the credentials?

Using related fields for lookups will never be airtight, but it's the status quo.

I don't like keeping unqualified name lookups, because it's a land mine for the user experience and unsupportable. Creating a "localhost" host will probably work once, and will certainly fail given time. If inventory is a hierarchal object, then why are organizations not?

We can add the lookup plugin and start pushing people towards it, but I don't see a way around using lookup fields like this. Eventually it can be generalized. My [WIP for that](https://github.com/ansible/awx/pull/6430) tried to generalize it _too much_, but we can converge on something, and in the meantime I'm in favor of adding some cases manually like this.

sean-m-sullivan commented:
My purpose for this was that the people I work with are planning on going to one tower, and the one thing holding them back was that job_template names could not be the same even if in different orgs, due to no related org. 

I believe their idea is they would have a dev org, which would hold dev playbooks and testing, and a prod ord which has the prod push branch templates and workflows. 

So the entire purpose of this was to search the org when looking up the name. I looked at how workflows did it, and did that solution. 

Do we want to remove the version lookup for now, and keep it as is similar to the workflow. and then once we land on the lookup, remove the code from both?

AlanCoding commented:
> Do we want to remove the version lookup for now, and keep it as is similar to the workflow. and then once we land on the lookup, remove the code from both?

Yes to removing the version lookup. Also, add to the README a note that this is backward-incompatible. Also, consider mentioning the backward-incompatibility in the documentation for the parameter. Correct docs never hurt.

Yes, I think it should work similar to workflow.

I don't see this functionality being removed. It could be refactored for all similar modules, that's all I was saying.

sean-m-sullivan commented:
So I've redone the code, and tested it vs 3.6.4, 3.7.0 and awx, It works in all 3, the 3.6 it works if you don't include the org, and don't have duplicate projects with the same name. I've included the following line in the documentation for the parameter
```
Requires Tower Version 3.7.0 or AWX 10.0.0 IS NOT backwards compatible with earlier versions.
```
I'm not sure where to put another part in the readme as you suggested.

AlanCoding commented on awx_collection/plugins/modules/tower_job_template.py:
```diff
@@ -45,6 +45,14 @@
+        - The Organization is inferred from the associated project
+        - If not provided, will lookup by name only, which does not work with duplicates.
+      type: str
+      version_added: 2.10
```
this version_added won't be accepted, it's complicated, we have to do an ignore rule for it. Might as well leave it off for now.

AlanCoding commented:
@sean-m-sullivan I think that text would be great to put in the documentation for the parameter. I think it's arguable that this _isn't_ backward incompatible since it doesn't cause problems if you don't use the parameter, so I'm fine to not put it in the README.md.

sean-m-sullivan committed:
updated to remove version and add docs

sean-m-sullivan committed:
updated version

sean-m-sullivan commented:
Updated to make requested changes Removed the version from the comment, added a note about backward compatibility in the parameter. Removed the version check from code.
